src/ctl/ctl.c

```diff
@@ -308,29 +389,36 @@ int ctl_query(struct ctl *ctl, void *ctx, enum ctl_query_source source,
      * easily retrieve the index values. The list is cleared once the ctl
      * query has been handled.
      */
-    struct ctl_index_utlist *indexes = NULL;
+    umf_ctl_index_utlist_t *indexes = NULL;
     indexes = Zalloc(sizeof(*indexes));
     if (!indexes) {
         return -1;
     }
 
     int ret = -1;
+    size_t name_offset = 0;
 
-    const struct ctl_node *n = ctl_find_node(CTL_NODE(global), name, indexes);
+    const umf_ctl_node_t *n =
+        ctl_find_node(CTL_NODE(global), name, indexes, &name_offset);
 
     if (n == NULL && ctl) {
         ctl_delete_indexes(indexes);
         indexes = NULL;
-        n = ctl_find_node(ctl->root, name, indexes);
+        n = ctl_find_node(ctl->root, name, indexes, &name_offset);
     }
 
-    if (n == NULL || n->type != CTL_NODE_LEAF || n->cb[type] == NULL) {
+    // if the appropriate node (leaf or subtree) is not found, then return error
+    if (n == NULL ||
+        (n->type != CTL_NODE_LEAF && n->type != CTL_NODE_SUBTREE) ||
+        n->cb[n->type == CTL_NODE_SUBTREE ? CTL_QUERY_SUBTREE : type] == NULL) {
         errno = EINVAL;
         goto out;
     }
 
-    ret = ctl_exec_query[type](ctx, n, source, arg, indexes);
-
+    const char *extra_name = (const char *)&name[0] + name_offset;
```

Comment:
with const char* do we still need casting?
`const char *extra_name = name + name_offset; `
wouldn't work? I'm just curious

Reply:
Done